zerver/lib/actions.py

```diff
@@ -3900,7 +3900,10 @@ def do_update_user_activity(user_profile_id: int,
         activity.save(update_fields=["last_visit", "count"])
 
 def send_presence_changed(user_profile: UserProfile, presence: UserPresence) -> None:
-    presence_dict = presence.to_dict()
+    # TODO: This needs to be rewritten for the new API.
+    # was previously presence.to_dict().
+    # May be an ugly transition since mobile handles these events.
+    presence_dict = {'client': 'website'}
```

Comment:
I think we need to do something with this TODO before we can try merging this PR.  Let me start a thread with mobile folks.

Reply:
Hmm, after a quick look, it's pretty clear that these will probably need to have a backwards-compatible format.  

@mateuszmandera can you work on writing the code to to that compatible reformatting, maybe in a `format_legacy_presence_dict` function in `zerver/lib/presence.py`?

Reply:
Yes, I wrote it based on my undestanding of what we want there?
```
def format_legacy_presence_dict(presence: UserPresence) -> Dict[str, Any]:
    """
    This function assumes it's being called right after the presence object was updated,
    and is not meant to be used on old presence data.
    """
    if presence.last_active_time + datetime.timedelta(minutes=1, seconds=10) >= presence.last_connected_time:
        status = UserPresence.LEGACY_STATUS_ACTIVE
        timestamp = datetime_to_timestamp(presence.last_active_time)
    else:
        status = UserPresence.LEGACY_STATUS_IDLE
        timestamp = datetime_to_timestamp(presence.last_connected_time)

    return dict(
        client='website',
        status=status,
        timestamp=timestamp,
        pushable=False
    )
```

The timedelta is selected based on old behavior
```
    stale_status = (log_time - presence.timestamp) > datetime.timedelta(minutes=1, seconds=10)
    was_idle = presence.status == UserPresence.IDLE
    became_online = (status == UserPresence.ACTIVE) and (stale_status or was_idle)
```
(which was determining when an `IDLE` update would "overpower" an older `ACTIVE` presence.)

Reply:
I'm also not sure if we want to be sending these legacy format event dicts to everyone for now the way I did it:
```
def send_presence_changed(user_profile: UserProfile, presence: UserPresence) -> None:
    # The mobile app handles these events so we need to use the old format.
    presence_dict = format_legacy_presence_dict(presence)
```

or somehow do this deeper in the event system to target older version of the mobile app only?

Reply:
I think we'll end up adding a `client_capabilities` option as part of migrating this API down the line; only older clients that don't declare it would get the old API.